### PR TITLE
format() returns the correct type

### PR DESCRIPTION
**Pull request by vladdeSV**
Changed `format()` to return the same type of string that was inputted.

For instance, formatting a *dstring* returns a *dstring* and not *string*.

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
Seems sane but might break existing code. Let's wait what others have to say

**Comment by yebblies** [comment]
This makes sense to me.

**Comment by vladdeSV** [comment]
@DmitryOlshansky Out of curiosity, can you come up with a scenario where this breaks code?

**Comment by vladdeSV** [comment]
I guess using `auto result = format(fmt, args)` and accidentally inserting a `[w|d]string` and still expect a `string`

**Comment by DmitryOlshansky** [comment]
> I guess using auto result = format(fmt, args) and accidentally inserting a [w|d]string and still expect a string

Yeah. This and maniacs using explicit types everywhere such as string for return and wstring for fmt. If it's going to hit simebody then it must be windows folks, being the most prominent use case for UTF-16.

**Comment by DmitryOlshansky** [comment]
Please squash the commits to one commit and it should be good to go

**Comment by DmitryOlshansky** [comment]
May need to whitelist @vladdeSV for auto-tester - @MartinNowak @braddr  ?

**Comment by vladdeSV** [comment]
Squashed the two commits. It should be good to go?

**Comment by vladdeSV** [comment]
One thing I think should be change is to replace my template name `C` with something like `String`

Thoughts?

**Comment by 9rnsr** [comment]
> One thing I think should be change is to replace my template name C with something like String

`C` is not good for "some sting" type. I'd recommend `S` to you.

**Comment by vladdeSV** [comment]
I also feel `S` is more suitable.

**Comment by vladdeSV** [comment]
I changed the `C` to `S`

**Comment by vladdeSV** [comment]
I changed `std.conv.convFormat` to fit `std.format.format`.

**However**, I do not see a reason to have `convFormat()`, as it's simply an alias for `format()`.

**Comment by vladdeSV** [comment]
This currently passes all tests, but I made some changes in `std.algorithm.iteration` where I cast a dstring to a string since it was used in an assert message @ 9097c7bc0b6d6d60bace40bb49c0eda10b4edbc0

Also, I'd say that c3e86ff4fcdd20b038b0665c2fd0a1bbaed12b13 should simply be an alias, if not to replace the function calls in the code to `std.format.format` completely.

**Comment by DmitryOlshansky** [comment]
One thing I really don't like is how it uses String as template parameter. Ideally it should accept const(Char)[] and produce mutable Char[]. Then if format is `pure` that mutable return is implicitly convertible to immutable(Char)[] and/or const(Char)[] at call site.


> This currently passes all tests, but I made some changes in std.algorithm.iteration where I cast a dstring to a string since it was used in an assert message @ 9097c7b

Using cast like that is very bad idea it just reinterprets the array, you should use `to!dstring(someStr)`.

Exactly the kind of thing to watch out for. We are early in release cycle so it might be fine to break a few cases like that.

File: std/conv.d [hunk 1]
```diff
@@ -27,7 +27,7 @@ import std.range.primitives;
 import std.traits;
 import std.typetuple;
 
-private string convFormat(Char, Args...)(in Char[] fmt, Args args)
+private S convFormat(S, Args...)(in S fmt, Args args) if (isSomeString!S)
```
Comment (DmitryOlshansky) [inline std/conv.d]: Do we even call it internally anywhere? It's private should be easy to just drop it...

**Comment by vladdeSV** [comment]
@DmitryOlshansky 
>Ideally it should accept const(Char)[] and produce mutable Char[]. Then if format is pure that mutable return is implicitly convertible to immutable(Char)[] and/or const(Char)[] at call site.

I see. Yes I agree, I'll do some testing.

>Using cast like that is very bad idea it just reinterprets the array, you should use to!dstring(someStr).

My bad, I did do `to!string([...])`

File: std/conv.d (hunk 1 above)
Comment (vladdeSV) [inline std/conv.d]: Yes, at a few places.
Line 1404, 1775, 3931 ...

**Comment by vladdeSV** [comment]
@DmitryOlshansky I tried, I failed. I don't know how to make the function as you described.

I tried to change `S` to `Char[]`, however I got errors in the manner of `cannot convert char[] to string` and I do now know how to make it passable. I reverted that commit.

**Comment by quickfur** [comment]
@vladdeSV Did you mark the function `pure`?

**Comment by quickfur** [comment]
ping @vladdeSV

**Comment by vladdeSV** [comment]
@quickfur Sorry, been busy past days. I'll try again :)

**Comment by vladdeSV** [comment]
Pong @quickfur This means I need to make `formattedWrite()` pure.

And a question: I built phobos with `make -f win32.mak` and I didn't get any errors about impure functions being called, but the auto tester does. Why?

**Comment by quickfur** [comment]
Most of Phobos is template code, so just building it may not necessarily reveal all compilation problems. Try `make -f win32.mak unittest` to also run unittests.

Usually that takes too long, though. I usually just run unittests for a specific module manually by doing `dmd -main -unittest -of/tmp/test std/format.d && /tmp/test` (or whatever the windows equivalent is).w

**Comment by vladdeSV** [comment]
Changing the function to be pure needs many changes in the code. For instance, I needed to change some things in `std.stdio` to be pure, just to make formatted write pure.

I don't really want to fiddle around in all of these files and just put pure on things I do not know anything about. Unless I get some help I'd rather make it return a correct string type instead of char type. But having it return correct type of char array is something I feel would be very nice to have.

Piong @quickfur

**Comment by DmitryOlshansky** [comment]
> Changing the function to be pure needs many changes in the code. For instance, I needed to change some things in std.stdio to be pure, just to make formatted write pure.

It's a a lot of work. Let's keep this simple and move with just changing the return type.

**Comment by vladdeSV** [comment]
@DmitryOlshansky Should I revert my latest commit?

**Comment by DmitryOlshansky** [comment]
@vladdeSV no, rather squash all of them to one

**Comment by vladdeSV** [comment]
@DmitryOlshansky Just to be clear: I should not keep the latest commit that fails the build?

**Comment by DmitryOlshansky** [comment]
@vladdeSV - kill pure, keep Char[], and make it one commit.

**Comment by vladdeSV** [comment]
@DmitryOlshansky I've squashed them to one, however just as before it does not build.

**Comment by DmitryOlshansky** [comment]
Well previously it was string aka immutable(char)[]
 So just make return type immutable(Char)[], later we may safely generalize it with purity.

**Comment by vladdeSV** [comment]
@DmitryOlshansky 
>So just make return type immutable(Char)[]

This causes problems when `return w.data;` is called, namely the same `Error: cannot implicitly convert expression (w.data()) of type char[] to string` as we got before.

I think the current best solution is to go back to my first suggestion. I can't make it return immutable (Char)[], but I can make it return a string of the correct type.

Any other suggestions?

File: std/format.d [hunk 2]
```diff
@@ -6540,11 +6545,11 @@ unittest
  * Params: fmt  = Format string. For detailed specification, see $(XREF _format,formattedWrite).
  *         args = Variadic list of arguments to format into returned string.
  */
-string format(Char, Args...)(in Char[] fmt, Args args)
+Char[] format(Char, Args...)(in Char[] fmt, Args args) if (isSomeChar!Char)
 {
     import std.format : formattedWrite, FormatException;
     import std.array : appender;
-    auto w = appender!string();
+    auto w = appender!(Char[]);
```
Comment (DmitryOlshansky) [inline std/format.d]: Char[] -- > immutable(Char)[]

File: std/format.d [hunk 3]
```diff
@@ -6540,11 +6545,11 @@ unittest
  * Params: fmt  = Format string. For detailed specification, see $(XREF _format,formattedWrite).
  *         args = Variadic list of arguments to format into returned string.
  */
-string format(Char, Args...)(in Char[] fmt, Args args)
+Char[] format(Char, Args...)(in Char[] fmt, Args args) if (isSomeChar!Char)
```
Comment (DmitryOlshansky) [inline std/format.d]: Char[] -- > immutable(Char)[]

**Comment by DmitryOlshansky** [comment]
I pointed specific places.

**Comment by vladdeSV** [comment]
Thank you @DmitryOlshansky !
The auto tester is doing it's stuff currently, but there should be no problems since I built it on my computer with no problems :+1:

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Pull request merged**